File: swift/problem226.swift
//Definition for a binary tree node.
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

//pre-order traversal
class Solution {
    func invertTree(_ root: TreeNode?) -> TreeNode? {
        let output = root
        var stack_node = [TreeNode?]()
        stack_node.append(output)
        while (stack_node.count > 0) {
            let temp = stack_node.popLast()
            let temp_node = temp??.left
            temp??.left = temp??.right
            temp??.right = temp_node
            if (temp??.left != nil) {
                stack_node.append(temp!!.left)
            }
            if (temp??.right != nil) {
                stack_node.append(temp!!.right)
            }
        }
        return output
    }
}

//mutual conversion between binary tree and array
class conversion {
    func a_to_b(_ value: [Int?]) -> TreeNode? {
        if (value[0] == nil) {
            return nil
        }
        
        var deque = [TreeNode?]()
        var index = 1
        let output = TreeNode(value[0]!)
        
        if (value.count == 1) {
            return output
        }
        
        deque.append(output)
        while (index < value.count) {
            let tmp = deque.removeFirst()
            if let val = value[index] {
                tmp?.left = TreeNode(val)
                deque.append(tmp?.left)
            }
            index += 1
            
            if let val = value[index] {
                tmp?.right = TreeNode(val)
                deque.append(tmp?.right)
            }
            index += 1
        }
        return output
    }

    func b_to_a(_ root: TreeNode?) -> [Int?] {
        if (root == nil) {
            return [nil as Int?]
        }
        var output = [Int?]()
        var deque = [TreeNode?]()
        
        deque.append(root)
        while (deque.count != 0) {
            let tmp = deque.removeFirst()
            if (tmp == nil) {
                output.append(nil)
            } else {
                output.append(tmp?.val)
                if (tmp?.left != nil || tmp?.right != nil) {
                    deque.append(tmp?.left)
                    deque.append(tmp?.right)
                }
            }
        }
        return output
    }
}

let test = conversion()
let input = [1,2,3,nil,4,5,6,7,nil,nil,8,9,10]
print("before inversion:", input)
print("after inversion:",test.b_to_a(Solution().invertTree(test.a_to_b(input))))

File: python/problem226.py
import time

''' 二叉树创建和遍历实现'''
class Node:
    def __init__(self,item):
        self.item = item
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    def add(self, item):
        node = Node(item)
        if self.root is None:
            self.root = node
        else:
            q = [self.root]

            while True:
                pop_node = q.pop(0)
                if pop_node.left is None:
                    pop_node.left = node
                    return
                elif pop_node.right is None:
                    pop_node.right = node
                    return
                else:
                    q.append(pop_node.left)
                    q.append(pop_node.right)

    def traverse(self):  # 层次遍历
        if self.root is None:
            return None
        q = [self.root]
        res = [self.root.item]
        while q != []:
            pop_node = q.pop(0)
            if pop_node.left is not None:
                q.append(pop_node.left)
                res.append(pop_node.left.item)

            if pop_node.right is not None:
                q.append(pop_node.right)
                res.append(pop_node.right.item)
        return res

    def preorder(self,root):  # 先序遍历
        if root is None:
            return []
        result = [root.item]
        left_item = self.preorder(root.left)
        right_item = self.preorder(root.right)
        return result + left_item + right_item

    def inorder(self,root):  # 中序遍历
        if root is None:
            return []
        result = [root.item]
        left_item = self.inorder(root.left)
        right_item = self.inorder(root.right)
        return left_item + result + right_item

    def postorder(self,root):  # 后序遍历
        if root is None:
            return []
        result = [root.item]
        left_item = self.postorder(root.left)
        right_item = self.postorder(root.right)
        return left_item + right_item + result
        
        '''二叉树结束'''
      
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.invert(root)
        return root
        
    def invert(self,node):
        if node and (node.right or node.left):
            node.left,node.right=node.right,node.left
            self.invert(node.right)
            self.invert(node.left)
            
    #from discuss       
    def invertTree2(self, root):
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                node.left, node.right = node.right, node.left
                stack += node.left, node.right
        return root
            
            
if __name__ == '__main__':
    t = Tree() 
    
    for i in range(8000):
        t.add(i)
    #time test for invertTree
    t1=time.time()
    c=Solution()
    c.invertTree(t.root)
    t2=time.time()
    print('时间为',t2-t1)
    
    #time test for invertTree2
    t3=time.time()
    c.invertTree2(t.root)
    t4=time.time()
    print('时间为',t4-t3)